tf/tfprocess.py

```diff
@@ -1135,61 +1197,122 @@ def residual_block(self, inputs, channels, name):
         # query, key, and value vectors for self-attention
-        q = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wq')(inputs)
-        k = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wk')(inputs)
-        v = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wv')(inputs)
+        # inputs b, 64, sz
+
+        q = tf.keras.layers.Dense(
+            d_model, name=name+'/wq', kernel_initializer='glorot_normal')(inputs)
+        k = tf.keras.layers.Dense(
+            d_model, name=name+'/wk', kernel_initializer='glorot_normal')(inputs)
+        v = tf.keras.layers.Dense(
+            d_model, name=name+'/wv', kernel_initializer=initializer)(inputs)
+
         # split q, k and v into smaller vectors of size 'depth' -- one for each head in multi-head attention
         batch_size = tf.shape(q)[0]
+
         q = self.split_heads(q, batch_size, num_heads, depth)
         k = self.split_heads(k, batch_size, num_heads, depth)
         v = self.split_heads(v, batch_size, num_heads, depth)
-        scaled_attention, attention_weights = self.scaled_dot_product_attention(q, k, v)
-        if num_heads > 1:
-            scaled_attention = tf.transpose(scaled_attention, perm=[0, 2, 1, 3])
-            scaled_attention = tf.reshape(scaled_attention, (batch_size, -1, d_model))  # concatenate heads
+
+        scaled_attention, attention_weights = self.scaled_dot_product_attention(
+            q, k, v, name=name, inputs=inputs)
+
         # final dense layer
-        output = tf.keras.layers.Dense(emb_size, kernel_initializer='glorot_normal',
-                                       name=name + "/dense")(scaled_attention)
+        output = tf.keras.layers.Dense(
+            emb_size, name=name + "/dense_fin", kernel_initializer=initializer)(scaled_attention)
         return output, attention_weights
-
+        
     # 2-layer dense feed-forward network in encoder blocks
-    def ffn(self, inputs, emb_size, dff, name):
-        dense1 = tf.keras.layers.Dense(dff, kernel_initializer='glorot_normal', activation='selu',
-                                       name=name + "/dense1")(inputs)
-        return tf.keras.layers.Dense(emb_size, kernel_initializer='glorot_normal', name=name + "/dense2")(dense1)
+    def ffn(self, inputs, emb_size: int, dff: int, initializer, name: str):
+        dense1 = tf.keras.layers.Dense(
+            dff, name=name + "/dense1", kernel_initializer=initializer)(inputs)
+
+        activation = square_relu if self.square_relu_ffn else tf.keras.activations.get(self.DEFAULT_ACTIVATION)
+        dense1 = activation(dense1)
+
+        out = tf.keras.layers.Dense(emb_size, name=name + "/dense2", kernel_initializer=initializer)(dense1)
+        return out
+
+    def encoder_layer(self, inputs, emb_size: int, d_model: int, num_heads: int, dff: int, name: str, training: bool):
+        # DeepNorm
+        alpha = tf.cast(tf.math.pow(
```

Comment:
Don't we need a path where alpha is forced to 1 for policy head encoder backwards compat for the non-attention body case?